### DISCUSSION TITLE
On the opportunity and ways to inject dependencies in advices

### DISCUSSION THREAD

**Discussion by picrap** [ArxOne/MrAdvice#179] [Ideas]
In issue #175, a discussion about DI was opened.
There was a [post](https://codereview.stackexchange.com/questions/20341/inject-dependency-into-postsharp-aspect) where options are shown.

The post above and referenced issues raise some questions:
- Is it a good idea to use DI within aspects?
- How to achieve this?

This is the time to see big, not to copy what others do.

**Comment by picrap** [comment]
I was thinking to steal properties from target, especially a `IServiceProvider`:
```csharp
public class SomeAdvice : Attribute, IMethodAdvice
{
    public void Advise(MethodAdviceContext context)
    {
        // -- some samples here:
        // find a value of type IServiceProvider in target
        var serviceProvider = context.GetMember<IServiceProvider>();
        // use a direct value or IServiceProvider
        var serviceProvider = context.GetMemberService<SomeService>();
        // the above is equivalent to
        var serviceProvider2 = (SomeService)context.GetMember<IServiceProvider>()?.GetService(typeof(SomeService))
                               ?? context.GetMember<SomeService>();

        // then proceed
        context.Proceed();
    }
}
```
Fun fact: this does not require any addition to MrAdvice core, but can be implemented as an extension class.

**Reply by arturgromek** [reply]
Hi, I've just started using this package and agree that a way to inject or use services such as for logging would be useful

**Comment by FrankLakmann** [comment]
Hello, thanks for your efforts.
We are currently migrating from old framework to .net 8. I would like to retrieve information about the current user from the app- serviceprovider inside our LoggingAspect : Attribute, IMethodAdvice.

> Fun fact: this does not require any addition to MrAdvice core, but can be implemented as an extension class.

I' not sure i understand that. What would this extension class look like to to make the 
`context.GetMember<IServiceProvider>();`
work in a LoggingAspect?

**Reply by picrap** [reply]
Since the advice has access to the target instance, it is easy to read its properties (or interfaces) and find the one you need.
The idea behind `context.GetMember<IServiceProvider>()` was to enumerate properties and use the one of type `IServiceProvider` (it is also very recommended to use a cache because properties enumeration is slow).

**Reply by FrankLakmann** [reply]
Ah ok, I will try and see if this is doable. The loggingaspect is perforemance sensitive for us an enumerating properties sounds expensive. Having an additional cache might help but complicates things.
Anyway, thanks again for the super quick response. Much appreciated.
